### PR TITLE
Move user config and repo config to Viper

### PR DESCRIPTION
**Pull request by gsoltis**
The main goal is break up the giant `Config` struct into pieces that will fit into a top-level helper / command base in the final move to `cobra`. 

 * Moves our top-level repo- and user-config file parsing to viper. 
 * Applies ENV VAR overrides as appropriate
 * Does not yet handle top-level flags in the mix, since that requires top-level `cobra`, but that is the eventual goal
 * Pulls out auth and endpoint details for the api client into a struct. This struct is generated with values from the repo and user configs

### CODE REVIEW

File: cli/internal/client/client.go [hunk 1]
```diff
@@ -40,17 +39,33 @@ type ApiClient struct {
 // ErrTooManyFailures is returned from remote cache API methods after `maxRemoteFailCount` errors have occurred
 var ErrTooManyFailures = errors.New("skipping HTTP Request, too many failures have occurred")
 
+// _maxRemoteFailCount is the number of failed requests before we stop trying to upload/download
+// artifacts to the remote cache
+const _maxRemoteFailCount = uint64(3)
```
Comment (mehulkar) [inline cli/internal/client/client.go:44]: I'm embarrassed to admit that I just learned that "unsigned int" has nothing to do with cryptography 😬

File: cli/internal/client/client.go [hunk 2]
```diff
@@ -40,17 +39,33 @@ type ApiClient struct {
...
+	c.token = token
+}
+
+// RemoteConfig holds the authentication and endpoint details for the API client
+type RemoteConfig struct {
+	Token    string
+	TeamID   string
+	TeamSlug string
+	APIURL   string
+}
```
Comment (mehulkar) [inline cli/internal/client/client.go:57]: 👏🏾  I like it

**Review by mehulkar — APPROVED** [review]
I had some trouble understanding the full scope of these changes and how they relate to each other (and I'm still getting comfortable with Go), but generally seems ok to me? 

Couple high level questions:

- Is viper the thing reading `turbo.json` or is this some other config?
- I noticed a bunch of changes in the tests. Are we keeping track of total coverage somehow to get a summary of whether it went up or down from a PR?

File: cli/internal/client/client.go [hunk 3]
```diff
@@ -23,10 +23,9 @@ import (
 type ApiClient struct {
 	// The api's base URL
 	baseUrl      string
-	Token        string
+	token        string
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply cli/internal/client/client.go:26]: It has the same effect as struct and function names. `.token` is now only visible in the `config` package. All of the other usages, including the tests, are in the `config` package.

[the remark this answers is not in the document]
Comment (mehulkar) [reply cli/internal/client/client.go:26]: Ah I see. I thought it would be private to the _struct_, rather than private to the package. Thanks!

**Pull request merged**